### PR TITLE
[devops] Linting and testing update

### PR DESCRIPTION
**Pull request by Noanan**
makefile and github actions to support linting, deadlock/race testing and coverage

- encountered issue with go work on windows not allowing go to find modules in /pkg/infra (might be windows only)
- we need to add tests for the deadlock/race issues and coveraging testing to come into effect

### CODE REVIEW

**Comment by 0x6770** [comment]
The "lint and format check" workflow works now, and has detected an unordered import of "github.com/benbjohnson/immutable" in `pkg/infra/main.go`.

Added a `.golangci.yaml` file to store configuration for `golangci-lint`

**Comment by antonypap** [comment]
Rebased the branch to remove the merge commit.
Not sure why the linter keeps on complaining about the immutable package import on (basically) every file

**Comment by antonypap** [comment]
Nice, was looking to format using goimports as well. This seems to fix it

**Comment by 0x6770** [comment]
> Nice, was looking to format using goimports as well. This seems to fix it

yep, but it seems like we still have a few formatting issues

**Comment by antonypap** [comment]
Fixing rest of the formatting issues rn so that we have a clean workspace with no errors

**Comment by 0x6770** [comment]
[`Go report card`](https://github.com/marketplace/actions/go-report-card) action does not run `goreportcard-cli`, instead, it is a wrapper to https://goreportcard.com. And it only check go entry file in the root.

If we want to run generate this report, we need to somehow install and run `goreportcard-cli` in workflow.

**Comment by 0x6770** [comment]
Although `goreportcard` action does not work for this project, most of its checks can be covered by `golangci-lint`, except `go vet` and license check, will add a new job to run `go vet`.

output from `goreportcard`
```
Grade .......... A+ 100.0%
Files ................. 18
Issues ................. 0
gofmt ............... 100%
go_vet .............. 100%
gocyclo ............. 100%
ineffassign ......... 100%
license ............. 100%
misspell ............ 100%
```

**Comment by antonypap** [comment]
Perfect, I think that this should be ready for review and merge now

File: Makefile [hunk 1]
```diff
@@ -18,3 +20,34 @@ runDebug: build
...
+		--enable=unconvert \
+		--enable=unparam \
+		--enable=asciicheck \
+		--enable=misspell \
+		--enable=decorder \
+		--enable=depguard \
+		--enable=nilerr \
+		--enable=exportloopref \
+		--enable=whitespace \
+		--enable=goimports
```
Comment (antonypap) [inline Makefile:43]: Could probably change this in the future to run with the [.golangci.yaml](.golangci.yaml) config file

**Review by antonypap — APPROVED** [review]
LGTM

File: Makefile (hunk 1 above)
Comment (0x6770) [reply Makefile]: ur right, I've updated Makefile

**Pull request merged**